pkg/apis/pipeline/v1alpha1/taskrun_validation.go

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2018 The Knative Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/knative/pkg/apis"
+	"k8s.io/apimachinery/pkg/api/equality"
+)
+
+func (t *TaskRun) Validate() *apis.FieldError {
+	if err := validateObjectMetadata(t.GetObjectMeta()).ViaField("metadata"); err != nil {
+		return err
+	}
+	return t.Spec.Validate()
+}
+
+func (ts *TaskRunSpec) Validate() *apis.FieldError {
+	if equality.Semantic.DeepEqual(ts, &TaskRunSpec{}) {
+		return apis.ErrMissingField("spec")
+	}
+
+	// Check for TaskRef
+	if ts.TaskRef.Name == "" {
+		return apis.ErrMissingField("spec.taskref.name")
+	}
```

Comment:
should we also validate that the referenced `Task` exists?

Reply:
For that kind of validation, pipeline client is required. In depth validation of related object existence could be done in taskrun controller.

Reply:
ah interesting, okay good to know!